### ISSUE TITLE
Throwing a TimeoutError in waitRead is too costly

### ISSUE THREAD

**Issue by vbargsten** [rock-core/drivers-iodrivers_base#45]
When checking for available data (e.g. through orogen component of iodrivers_base hasIO()) or reading until no more data is available, throwing and catching exceptions is not a good approach as it is computationally costly (via unwind etc.). An alternative method should be discussed.

Exemplary perf output:
![grafik](https://user-images.githubusercontent.com/9880945/134697971-165893ba-78ba-4b42-81f2-be1d8db0ea39.png)

**Comment by vbargsten** [comment]
I can't quantify it in ms. In the example, the updatehook takes 6% of the (i7 10th gen) CPU, and from that ca. two third is just the exception unwind or roughly 90% of the serial IO processing (incl. decoding, encoding, crc checking etc.). It is running with 7 serial channels (two FT4232H), connected to 30 devices, 24 of them with 1 kHz sampling rate. I tried to fix that without changing the function definitions by introducing  `silent_timeout` flag, but that's not a very clean solution (https://github.com/vbargsten/drivers-iodrivers_base/commit/fea02805c477841e771257f6d57b1d281c69e6b0).

**Comment by g-arjones** [comment]
I think it makes perfect sense to refractor IOStream::waitRead into a IOStream::hasIO() method that will just poll the fd and call that in iodriver_base::hasIO.

That said, I don't think you will see a great improvement in performance.

Also, I think your patch changes/breaks the component's behavior

**Comment by vbargsten** [comment]
> I think it makes perfect sense to refractor IOStream::waitRead into a IOStream::hasIO() method that will just poll the fd and call that in iodriver_base::hasIO.

It is not only hasIO, but the read itself as well if we read until no more data is available. The waitRead function makes sense in my eyes, but it could e.g. return bool instead of throwing an exception. I added a flag, so the behavior can at least be configured.

> That said, I don't think you will see a great improvement in performance.

Regarding the performance: Well, in absolute amount the difference on a powerful system is only a few percent, the relative reduction in my case is around 70 %. Such "small" components should not be that wasteful. 

> Also, I think your patch changes/breaks the component's behavior

You're right. I forgot to push that one https://github.com/vbargsten/drivers-orogen-iodrivers_base/commit/844ce806e238091b831d945341ad47d5225a3faf. I added a property to set this flag in my component, so I can compare easily.

**Comment by vbargsten** [comment]
I agree for readPacket, but what about readRaw? There must be some option to just read whatever is available at the moment without penalties (I also try to avoid select for serial channels at least, just calling read is usually faster).

Anyways...collecting here the changes:
#46 
https://github.com/rock-core/drivers-orogen-iodrivers_base/pull/31

Will test next week.

**Comment by doudou** [comment]
> but what about readRaw?

Same. **Especially** with readRaw which has no issue with packet handling.

> (I also try to avoid select for serial channels at least, just calling read is usually faster).

This is what readPacket does: read first, and then call waitRead to wait if needed (you can't read with a timeout).

The while(hasIO()) ... loop could be optimized by having a prefetch() instead which reads and fills the internal buffer, and returns true/false, but I'm really not sure the optimization would be worth it (and even if you do want to implement it, please keep it separate from the current change).

There is other room for optimizations (for instance extractPacket is called too many times), but I don't think the complexity in the resulting API would be worth the optimization.

**Comment by doudou** [comment]
> Same. **Especially** with readRaw which has no issue with packet handling.

OK ... Now that I actually re-read the code I get your question. Sorry for the hasty answer.

readRaw already should not be  throwing if there is no data (documentation says "This never throws", which might actually not be true if it throws on I/O error). I guess it could be refactored to avoid waitRead if the global timeout is zero.

**Comment by vbargsten** [comment]
With the current changes, reduction is around 80 to 90 %. 

Most expensive (but on a much smaller scale) is select now. I'm not sure why lib_c_read is not visible anymore (also using the search) and why select is not visible as callee of the waitRead through readRaw but only through hasIO, though.

Exemplary perf log:
![grafik](https://user-images.githubusercontent.com/9880945/135844674-edb191fc-9d0b-445b-a5fd-f5eb2b47dca8.png)

**Comment by vbargsten** [comment]
> how much CPU is it taking now ?

It fell very roughly from around 5.7..6% to 0.7...1%. I would not take these numbers of CPU usage share of the full system too serious; it is more important to compare it to the related processes (e.g. we saw initially the unwind took much more compared to the actual read and processing).

> Isn't there a proper packetization that would allow to use readPacket() instead ?

readPacket will also call waitRead etc. with the additional requirement to get a packet or throw.

>  switching to poll()

Tried that 2 years ago* and I can not confirm that it is faster or has less latency for a single file descriptor. A big problem is also that it takes a timeout value in millisecond resolution. (see e.g. https://vmsplice.net/~stefan/stefanha-fosdem-2021.pdf p. 30). * on an embedded system; the fastest was to just use read and sleep.

So I my eyes it is fine to use select here, only if timeout is zero we could skip it and read directly.

**Comment by doudou** [comment]
Fixed by #46
